gerber/lib/Gerber/EncodedDecimal.hs

```diff
@@ -1,9 +1,53 @@
-module Gerber.EncodedDecimal ( EncodedDecimal(..) ) where
+{-# language DeriveGeneric #-}
+{-# language DeriveAnyClass #-}
+{-# language PatternSynonyms #-}
+{-# language ViewPatterns #-}
+{-# language BangPatterns #-}
+{-# language FlexibleContexts #-}
+{-# language NamedFieldPuns #-}
+
+module Gerber.EncodedDecimal ( EncodedDecimal(EncodedDecimal, StoredEncodedDecimal), negative, digits ) where
+
+import Data.Int
+import Data.Word
+
+pattern EncodedDecimal :: Bool -> [ Int ] -> EncodedDecimal
+pattern EncodedDecimal {negative, digits} <- (toEncoded -> (negative, digits)) where
+  EncodedDecimal n d = fromEncoded n d
+
+{-# COMPLETE EncodedDecimal #-}
 
 data EncodedDecimal =
-  EncodedDecimal
-    { negative :: Bool
-    , digits :: [ Int ]
+  StoredEncodedDecimal
+    { width :: {-# UNPACK #-} !Word8
+    , digitsAsInt :: {-# UNPACK #-} !Int32
     }
   deriving
-    ( Eq, Show )
+    ( Eq )
+
+instance Show EncodedDecimal where
+   show EncodedDecimal{negative = n, digits = d} =
+    "EncodedDecimal {negative = " ++ show n ++ ", digits = " ++ show d ++ "}"
+
+
+toEncoded :: EncodedDecimal -> (Bool, [ Int ])
+toEncoded StoredEncodedDecimal {width, digitsAsInt} = (digitsAsInt < 0, reverse (go 0 (fromIntegral (abs digitsAsInt))))
+  where
+    go :: Word8 -> Int -> [Int]
+    go !currentWidth !num
+      | currentWidth == width = []
+      | otherwise = num `mod` 10 : go (currentWidth + 1) (num `div` 10)
+
+fromEncoded :: Bool -> [Int] -> EncodedDecimal
+fromEncoded isNegative digitList =
+  StoredEncodedDecimal
+    { width = fromIntegral (length digitList)
+    , digitsAsInt = fromIntegral . doNegation . go 1 0 . reverse $ digitList
+    }
+  where
+    go _ !sum_ [] = sum_
+    go !magnitude !sum_ (digit:digits_) = go (magnitude * 10) (sum_ + magnitude * digit) digits_
+
+    doNegation
+      | isNegative = negate
+      | otherwise = id
```

Comment:
Just to play a little bit of :golf:, we could remove the explicit recursion and use `foldl`. Up to you if you think this is any better! Disclaimer: I haven't type checked this, so maybe I got it wrong and made things worse!

```suggestion
fromEncoded :: Bool -> [Int] -> EncodedDecimal
fromEncoded isNegative = doNegation . foldl' step zero . reverse
  where
    zero = StoredEncodedDecimal 0 0

    step StoredEncodedDecimal{ width, digitsAsInt } digit = StoredEncodedDecimal 
      { width = width + 1
      , digitsAsInt = digitsAsInt + fromIntegral digit * 10 ^ width
      }

    doNegation
      | isNegative = negate
      | otherwise = id
```